dcrec/secp256k1/signature.go

```diff
@@ -392,171 +409,275 @@ func canonicalPadding(b []byte) error {
+
+	// compactSigMagicOffset is a value used when creating the compact signature
+	// recovery code inherited from Bitcoin and has no meaning, but has been
+	// retained for compatibility.  For historical purposes, it was originally
+	// picked to avoid a binary representation that would allow compact
+	// signatures to be mistaken for other components.
+	compactSigMagicOffset = 27
+
+	// compactSigCompPubKey is a value used when creating the compact signature
+	// recovery code to indicate the original public key was compressed.
+	compactSigCompPubKey = 4
+
+	// pubKeyRecoveryCodeOddnessBit specifies the bit that indicates the oddess
+	// of the Y coordinate of the random point calculated when creating a
+	// signature.
+	pubKeyRecoveryCodeOddnessBit = 1 << 0
+
+	// pubKeyRecoveryCodeOverflowBit specifies the bit that indicates the X
+	// coordinate of the random point calculated when creating a signature was
+	// >= N, where N is the order of the group.
+	pubKeyRecoveryCodeOverflowBit = 1 << 1
+)
 
-// recoverKeyFromSignature recovers a public key from the signature "sig" on the
-// given message hash "msg". Based on the algorithm found in section 5.1.5 of
-// SEC 1 Ver 2.0, page 47-48 (53 and 54 in the pdf). This performs the details
-// in the inner loop in Step 1. The counter provided is actually the j parameter
-// of the loop * 2 - on the first iteration of j we do the R case, else the -R
-// case in step 1.6. This counter is used in the Decred compressed signature
-// format and thus we match bitcoind's behaviour here.
-func recoverKeyFromSignature(sig *Signature, msg []byte, iter int, doChecks bool) (*PublicKey, error) {
-	// 1.1 x = (n * i) + r
-	curve := S256()
-	Rx := new(big.Int).Mul(curve.Params().N,
-		new(big.Int).SetInt64(int64(iter/2)))
-	Rx.Add(Rx, sig.r)
-	if Rx.Cmp(curve.Params().P) != -1 {
-		return nil, errors.New("calculated Rx is larger than curve P")
-	}
-
-	// convert 02<Rx> to point R. (step 1.2 and 1.3). If we are on an odd
-	// iteration then 1.6 will be done with -R, so we calculate the other
-	// term when uncompressing the point.
-	Ry, err := decompressPoint(Rx, iter%2 == 1)
-	if err != nil {
-		return nil, err
-	}
-
-	// 1.4 Check n*R is point at infinity
-	if doChecks {
-		nRx, nRy := curve.ScalarMult(Rx, Ry, curve.Params().N.Bytes())
-		if nRx.Sign() != 0 || nRy.Sign() != 0 {
-			return nil, errors.New("n*R does not equal the point at infinity")
-		}
+// SignCompact produces a compact signature of the data in hash with the given
+// private key on the secp256k1 curve. The isCompressedKey parameter specifies
+// if the given signature should reference a compressed public key or not.
+//
+// Compact signature format:
+// <1-byte compact sig recovery code><32-byte R><32-byte S>
+//
+// The compact sig recovery code is the value 27 + public key recovery code + 4
+// if the compact signature was created with a compressed public key.
+func SignCompact(key *PrivateKey, hash []byte, isCompressedKey bool) []byte {
```

Comment:
is there any reason to allow uncompressed keys?

Reply:
Compact signatures aren't enforced by consensus so there could be some in the wild, so I didn't think we should remove support.

Reply:
discussed in chat and I don't agree that we should be providing features that decred software doesn't need and shouldn't use in our packages, but it's not something to get hung up on to block this pr.